DataStructures/Timeline.cs

```diff
@@ -0,0 +1,547 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace DataStructures
+{
+    /// <summary>
+    /// A collection of <see cref="DateTime"/> and <see cref="TValue"/> sorted by <see cref="DateTime"/> field.
+    /// </summary>
+    /// <typeparam name="TValue">Value associated with a <see cref="DateTime"/>.</typeparam>
+    public class Timeline<TValue>:
+        IEnumerable<(DateTime Time, TValue Value)>,
+        ICollection<(DateTime Time, TValue Value)>,
+        IEquatable<Timeline<TValue>>
+    {
+        private List<(DateTime Time, TValue Value)> timeline;
+        bool ICollection<(DateTime Time, TValue Value)>.IsReadOnly => false;
+
+        /// <summary>
+        /// The count of pairs.
+        /// </summary>
+        public int Count => timeline.Count;
+
+        /// <summary>
+        /// The count of times.
+        /// </summary>
+        public int TimesCount => GetAllTimes().Length;
+
+        /// <summary>
+        /// The count of values.
+        /// </summary>
+        public int ValuesCount => GetAllValues().Length;
+
+        /// <summary>
+        /// Get all values associated with <paramref name="time"/>.
+        /// </summary>
+        /// <returns>Values associated with <paramref name="time"/>.</returns>
+        public TValue[] this[DateTime time]
+        {
+            get => GetValuesByTime(time);
+            set
+            {
+                var c = -1;
+                int i;
+                for (i = 0; i < Count; i++)
+                {
+                    if (timeline[i].Time == time)
+                    {
+                        if (c == value.Length)
+                            timeline.RemoveAt(i);
+                        else
+                            timeline[i] = (timeline[i].Time, value[++c]);
+                    }
+                }
+                while (i < c)
+                    Add(time, value[c]);
```

Comment:
Could you comment this piece of code a bit because I don't completely get an idea of what's going on